### PR TITLE
Migrate to oneDNN 3.0 API

### PR DESCRIPTION
**Pull request by Xia-Weiwen**
Migrate to oneDNN 3.0 API for IPEX and PyTorch.

**Major changes of oneDNN**
- `dnnl::memory::desc` struct become opaque. Can only access via API. Read only, cannot write.
- `primitive_attr` no longer keeps info of scales or zero points. Scales and zero points are set at runtime. Set masks of scales and zero points to `primitive_attr` to create primitive desc (pd).
- Output scales are no longer used for pd creation and computation. Set src/wei/dst, etc. scales separately.
- API to create `primitive_desc` is changed
- Parameter `scale` is removed from post-op fusion API. Some default parameters are changed (e.g., `hardswish`).
- Scale and shift are separated for batchnorm
- A few deprecated APIs are removed.

**Changes in ideep**
- `dnnl::primitive_attr` no longer holds scales and zero points.
  - Keep scales and zero points as well as their masks in `ideep::attr_t`. Define `std::unordered_map<int, scale-mask-pair>` and `std::unordered_map<int, zp-mask-pair>` in `attr_t`. Keys are DNNL args (e.g. `DNNL_ARG_SRC`)
  - Keep API of `set_scales` and `set_zero_points` of `ideep::attr_t`. They set masks to base class `dnnl::primitive_attr` and store these info in `ideep::attr_t::all_scales` and `ideep::attr_t::all_zero_points`, which are maps mentioned above.
  - Removed `forward_quant_parameters` for conv and deconv which used to keep scale and zp. Define `std::shared_ptr<std::unordered_map<int, tensor>> all_scales, all_zero_points` in `forward_params` to keep scales and zps as tensors for primitive execution. They are filled in `do_prepare`.
  - For quantization, fp32 bias does not need to be scaled before passing to primitive.
- `dnnl::memory::desc` has become opaque.
  - `ideep::tensor::desc::to_type()` returns a new desc by cloning an existing one and change its data type. Now we cannot change a desc's data type directly. Solution: (1) Use strides and data type to create a new desc if the given desc is in plain format. (2) Create a binary primitive desc to query a new desc if the given desc is in blocked format.
  - Keep group info in `ideep::tensor::desc` and `ideep::tensor`. We used to keep it in `dnnl::memory::desc`.
- API to create `primitive_desc` is changed.
  - Use new API to create `primitive_desc` for various ops.
- Post-op fusion API is changed. Parameter `scale` is removed. 
  - Remove the parameter in ideep. For some post-ops, old APIs are used with scales (`1.0f`). we need to keep the old API and name the new as `_v2`

**Test Plan**
- IPEX UT
- Some of PyTorch UT (test_quantization.py, test_mkldnn.py, test_mkldnn_fusion.py)

@jgong5 @XiaobingSuper @yanbing-j @Guobing-Chen @EikanWang Please review. Thanks!

### CODE REVIEW

File: include/ideep/operators/deconv.hpp [hunk 1]
```diff
@@ -670,15 +663,34 @@ struct convolution_transpose_forward : public dnnl::deconvolution_forward {
...
-    param.sq_param_ptr =
-        std::make_shared<deconv_forward_quant_params>(std::move(src_zero_point_m));
-    IDEEP_ENFORCE(param.sq_param_ptr, "Failed to allocate memory for quantization parameters");
+
+    // Prepare tensors for scales and zero points
+    if (param.op_attr.has_scales()) {
+      if (!param.all_scales) {
+        param.all_scales.reset(new std::unordered_map<int, tensor>);
+      }
+      for (auto& arg_scale_pair : param.op_attr.get_all_scales()) {
```
Comment (jgong5) [inline include/ideep/operators/deconv.hpp:676]: Why can't we use `param.op_attr.get_all_scales()` directly in the execution?

File: include/ideep/operators/inner_product.hpp [hunk 2]
```diff
@@ -560,17 +558,6 @@ struct inner_product_backward_data : public dnnl::inner_product_backward_data {
...
-      if (weights_.get_data_type() != data_type::bf16) {
-        weights_.init(weights.get_desc().to_type(data_type::bf16));
-        weights_.reorder_from(weights);
-      }
-    } else if (diff_dst.get_data_type() == data_type::f16) {
-      if (weights_.get_data_type() != data_type::f16) {
-        weights_.init(weights.get_desc().to_type(data_type::f16));
-        weights_.reorder_from(weights);
-      }
-    }
```
Comment (jgong5) [inline include/ideep/operators/inner_product.hpp:573]: These lines can be safely deleted?

File: include/ideep/tensor.hpp [hunk 3]
```diff
@@ -285,9 +302,25 @@ class tensor : public memory {
...
+      }
+      // Case 3: For blocked layout, we cannot create a new desc with strides
+      // Use binary primitive desc to query a desc so that we can
+      // change the data type and keep the layout
+      auto& src0_desc = *this;
+      auto src1_desc = desc(get_internal_dims(), atype, tag::any);
+      auto dst_desc = desc(get_internal_dims(), atype, tag::any);
+      auto pd = dnnl::binary::primitive_desc(
+          aengine, algorithm::binary_add, src0_desc, src1_desc, dst_desc);
+      auto ret = desc(pd.dst_desc());
```
Comment (jgong5) [inline include/ideep/tensor.hpp:323]: This is with the assumption that `binary_add` keeps the blocked layout. It might be always true but seems to be something depending on implementation details. Is it what recommended by oneDNN folks?

File: include/ideep/operators/deconv.hpp (hunk 1 above)
Comment (Xia-Weiwen) [reply include/ideep/operators/deconv.hpp:676]: > Why can't we use `param.op_attr.get_all_scales()` directly in the execution?

Because `op_attr` stores scales and zero points as `vector`s but execution needs `tensor`s. So, here we construct `tensor`s and put them in `forward_params`.

File: include/ideep/operators/inner_product.hpp (hunk 2 above)
Comment (Xia-Weiwen) [reply include/ideep/operators/inner_product.hpp:573]: > These lines can be safely deleted?

I think so. The purpose of these lines is to get a weight tensor in `f16` data type and use its desc for pd creation. But actually, we don't need to do reorder here. We can simply create a weight desc by shape, data type, and `tag::any` for pd creation (It's a suggestion from onednn team when we were discussing the `to_type` issue).
The weight tensor will then be compared with expected desc (queried from pd) and reorder if needed.

File: include/ideep/tensor.hpp (hunk 3 above)
Comment (Xia-Weiwen) [reply include/ideep/tensor.hpp:323]: > This is with the assumption that `binary_add` keeps the blocked layout. It might be always true but seems to be something depending on implementation details. Is it what recommended by oneDNN folks?

Yes. We (mainly Haozhe) have raised our concern. OneDNN team admitted the potential risk, but they said they would not change such behavior and suggested we use it with confidence.

File: include/ideep/operators/deconv.hpp (hunk 1 above)
Comment (jgong5) [reply include/ideep/operators/deconv.hpp:676]: Would there be copy overhead for per-channel quant scales?

Comment (Xia-Weiwen) [reply include/ideep/operators/deconv.hpp:676]: Yes, if we want to keep the `const` semantics. It happens in preparation stage, so it does not matter if we have cache. The copying seems inevitable, and I haven't thought of a good idea yet.

**Comment by Xia-Weiwen** [comment]
Hi @jgong5. May I know if this PR is approved? Can we merge it? If not, will you (and probably others) review here or have a review meeting? Thanks!

**Review by jgong5 — APPROVED** [review]
